File: gitwatcher/services/github.py
from urllib.parse import urlparse, quote
import httpx
from ..database import repositories as repo_db
from .encryption import decrypt_secret

API = "https://api.github.com"
API_VERSION = "2022-11-28"

def _headers(token=None):
    headers = {
        "Accept": "application/vnd.github+json",
        "X-GitHub-Api-Version": API_VERSION,
        "User-Agent": "GitWatcher-Discord-Bot",
    }
    if token:
        headers["Authorization"] = f"Bearer {token}"
    return headers

def parse_github_profile(value: str) -> str:
    value = value.strip()
    if "://" not in value:
        value = value.lstrip("@")
        if not value or "/" in value:
            raise ValueError("Enter a GitHub username or profile URL.")
        return value
    parsed = urlparse(value)
    if parsed.scheme not in ("http", "https") or parsed.netloc.lower() not in ("github.com", "www.github.com"):
        raise ValueError("That is not a GitHub profile URL.")
    parts = [p for p in parsed.path.strip("/").split("/") if p]
    if len(parts) != 1:
        raise ValueError("Paste a GitHub profile URL, not a repository URL.")
    return parts[0]

def parse_repository_url(value: str):
    parsed = urlparse(value.strip())
    if parsed.scheme not in ("http", "https") or parsed.netloc.lower() not in ("github.com", "www.github.com"):
        raise ValueError("Paste a normal GitHub repository URL.")
    parts = [p for p in parsed.path.strip("/").split("/") if p]
    if len(parts) < 2:
        raise ValueError("That does not look like a GitHub repository URL.")
    owner, repo = parts[0], parts[1]
    if repo.endswith(".git"):
        repo = repo[:-4]
    return owner, repo

def guild_token(guild_id):
    credential = repo_db.get_github_credential(guild_id)
    return decrypt_secret(credential["token_encrypted"]) if credential else None

async def get_user(username, token=None):
    async with httpx.AsyncClient(timeout=15) as client:
        r = await client.get(f"{API}/users/{quote(username, safe='')}", headers=_headers(token))
    if r.status_code == 404:
        raise ValueError(f"GitHub user `{username}` does not exist.")
    r.raise_for_status()
    return r.json()

async def get_authenticated_user(token):
    async with httpx.AsyncClient(timeout=15) as client:
        r = await client.get(f"{API}/user", headers=_headers(token))
    if r.status_code == 401:
        raise ValueError("GitHub rejected that token.")
    r.raise_for_status()
    return r.json()

async def validate_repository(guild_id, owner, repo):
    token = guild_token(guild_id)
    async with httpx.AsyncClient(timeout=15) as client:
        r = await client.get(f"{API}/repos/{quote(owner, safe='')}/{quote(repo, safe='')}", headers=_headers(token))
        if r.status_code == 404:
            raise ValueError("I cannot find that repository. If it is private, run `/gitwatcher auth` first.")
        if r.status_code == 403:
            raise ValueError("GitHub denied access to that repository.")
        r.raise_for_status()
        info = r.json()
        b = await client.get(f"{API}/repos/{quote(owner, safe='')}/{quote(repo, safe='')}/branches/main", headers=_headers(token))
        if b.status_code == 404:
            raise ValueError("GitWatcher MVP only watches a branch named `main`.")
        b.raise_for_status()
        branch = b.json()
    return {
        "owner": info["owner"]["login"],
        "repo": info["name"],
        "is_private": bool(info.get("private")),
        "head_sha": branch["commit"]["sha"],
    }

async def current_main_head(repository):
    token = guild_token(repository["guild_id"])
    async with httpx.AsyncClient(timeout=15) as client:
        r = await client.get(
            f"{API}/repos/{quote(repository['owner'], safe='')}/{quote(repository['repo'], safe='')}/branches/main",
            headers=_headers(token),
        )
    r.raise_for_status()
    return r.json()["commit"]["sha"]

async def compare_commits(repository, base_sha, head_sha):
    token = guild_token(repository["guild_id"])
    owner = quote(repository["owner"], safe="")
    repo = quote(repository["repo"], safe="")
    basehead = quote(f"{base_sha}...{head_sha}", safe=".")
    commits = []
    page = 1
    async with httpx.AsyncClient(timeout=20) as client:
        while True:
            r = await client.get(
                f"{API}/repos/{owner}/{repo}/compare/{basehead}",
                params={"per_page": 100, "page": page},
                headers=_headers(token),
            )
            if r.status_code in (404, 409):
                raise RuntimeError("Git history changed and the previous checkpoint can no longer be compared.")
            r.raise_for_status()
            payload = r.json()
            if payload.get("status") not in ("ahead", "identical"):
                raise RuntimeError("The branch history is no longer a simple continuation of the saved checkpoint.")
            batch = payload.get("commits") or []
            commits.extend(batch)
            if len(batch) < 100:
                break
            page += 1
            if page > 20:
                raise RuntimeError("Too many commits to process safely in one check.")
    return commits

File: src/services/github.js
import axios from 'axios';
import { pool } from '../db/pool.js';
import {
  decryptSecret,
  encryptSecret,
  hasDedicatedEncryptionKey,
  isLegacySecret,
} from './crypto.js';

const API = 'https://api.github.com';

function headers(token) {
  const value = {
    Accept: 'application/vnd.github+json',
    'X-GitHub-Api-Version': '2022-11-28',
    'User-Agent': 'GitWatcher-Discord-Bot',
  };

  if (token) value.Authorization = `Bearer ${token}`;
  return value;
}

export function parseProfile(value) {
  const trimmed = value.trim();

  let username = trimmed;

  if (trimmed.includes('://')) {
    const url = new URL(trimmed);

    if (
      !['github.com', 'www.github.com'].includes(url.hostname.toLowerCase()) ||
      url.username ||
      url.password ||
      url.port ||
      url.search ||
      url.hash
    ) {
      throw new Error('That is not a normal GitHub profile URL.');
    }

    const parts = url.pathname.split('/').filter(Boolean);
    if (parts.length !== 1) {
      throw new Error('Paste a GitHub profile URL, not a repository URL.');
    }

    username = parts[0];
  } else {
    username = trimmed.replace(/^@/, '');
  }

  if (
    !username ||
    username.length > 39 ||
    !/^[A-Za-z0-9](?:[A-Za-z0-9-]{0,37}[A-Za-z0-9])?$/.test(username)
  ) {
    throw new Error('Enter a valid GitHub username or profile URL.');
  }

  return username;
}

export function parseRepo(value) {
  let url;

  try {
    url = new URL(value.trim());
  } catch {
    throw new Error('Paste a full GitHub repository URL beginning with https://github.com/.');
  }

  if (
    url.protocol !== 'https:' ||
    !['github.com', 'www.github.com'].includes(url.hostname.toLowerCase()) ||
    url.username ||
    url.password ||
    url.port ||
    url.search ||
    url.hash
  ) {
    throw new Error('Paste a normal HTTPS GitHub repository URL.');
  }

  const parts = url.pathname.split('/').filter(Boolean);
  if (parts.length !== 2) {
    throw new Error('Paste the repository root URL, for example https://github.com/owner/repo.');
  }

  const owner = parts[0];
  const repo = parts[1].replace(/\.git$/i, '');

  if (
    !/^[A-Za-z0-9](?:[A-Za-z0-9-]{0,37}[A-Za-z0-9])?$/.test(owner) ||
    !repo ||
    repo.length > 100 ||
    !/^[A-Za-z0-9._-]+$/.test(repo)
  ) {
    throw new Error('That GitHub owner or repository name is invalid.');
  }

  return { owner, repo };
}

export async function guildToken(guildId) {
  const { rows } = await pool.query(
    'SELECT token_encrypted FROM github_credentials WHERE guild_id=$1',
    [guildId],
  );

  if (!rows[0]) return null;

  const stored = rows[0].token_encrypted;
  const token = decryptSecret(stored);

  // Seamlessly migrate old Discord-token-derived ciphertext once a dedicated
  // Railway encryption key is configured.
  if (isLegacySecret(stored) && hasDedicatedEncryptionKey()) {
    await pool.query(
      'UPDATE github_credentials SET token_encrypted=$1, updated_at=NOW() WHERE guild_id=$2',
      [encryptSecret(token), guildId],
    );
  }

  return token;
}

export async function getUser(username, token = null) {
  try {
    const response = await axios.get(
      `${API}/users/${encodeURIComponent(username)}`,
      { headers: headers(token), timeout: 15000 },
    );
    return response.data;
  } catch (error) {
    if (error.response?.status === 404) {
      throw new Error(`GitHub user "${username}" does not exist.`);
    }
    throw error;
  }
}

export async function authenticatedUser(token) {
  try {
    const response = await axios.get(`${API}/user`, {
      headers: headers(token),
      timeout: 15000,
    });
    return response.data;
  } catch (error) {
    if (error.response?.status === 401) {
      throw new Error('GitHub rejected that token.');
    }
    throw error;
  }
}

export async function validateRepo(guildId, owner, repo) {
  const token = await guildToken(guildId);

  try {
    const [repoResponse, branchResponse] = await Promise.all([
      axios.get(`${API}/repos/${encodeURIComponent(owner)}/${encodeURIComponent(repo)}`, {
        headers: headers(token),
        timeout: 15000,
      }),
      axios.get(
        `${API}/repos/${encodeURIComponent(owner)}/${encodeURIComponent(repo)}/branches/main`,
        { headers: headers(token), timeout: 15000 },
      ),
    ]);

    return {
      owner: repoResponse.data.owner.login,
      repo: repoResponse.data.name,
      isPrivate: Boolean(repoResponse.data.private),
      headSha: branchResponse.data.commit.sha,
    };
  } catch (error) {
    if (error.response?.status === 404) {
      throw new Error(
        token
          ? 'I cannot find that repo with this server’s GitHub access.'
          : 'I cannot find that repo. If it is private, run /gitwatcher auth first.',
      );
    }
    throw error;
  }
}

export async function mainHead(repository) {
  const token = await guildToken(repository.guild_id);

  const response = await axios.get(
    `${API}/repos/${encodeURIComponent(repository.owner)}/${encodeURIComponent(repository.repo)}/branches/main`,
    { headers: headers(token), timeout: 15000 },
  );

  return response.data.commit.sha;
}

export async function compare(repository, base, head) {
  const token = await guildToken(repository.guild_id);
  const response = await axios.get(
    `${API}/repos/${encodeURIComponent(repository.owner)}/${encodeURIComponent(repository.repo)}/compare/${base}...${head}`,
    {
      headers: headers(token),
      timeout: 20000,
      params: { per_page: 100 },
    },
  );

  return response.data.commits || [];
}


export async function branchHead(guildId, owner, repo, branch) {
  const token = await guildToken(guildId);

  try {
    const response = await axios.get(
      `${API}/repos/${encodeURIComponent(owner)}/${encodeURIComponent(repo)}/branches/${encodeURIComponent(branch)}`,
      {
        headers: headers(token),
        timeout: 15000,
      },
    );

    return {
      exists: true,
      sha: response.data.commit.sha,
      branch: response.data.name,
    };
  } catch (error) {
    if (error.response?.status === 404) {
      return { exists: false };
    }
    throw error;
  }
}

export async function compareBranchRange(guildId, owner, repo, base, head) {
  const token = await guildToken(guildId);

  const response = await axios.get(
    `${API}/repos/${encodeURIComponent(owner)}/${encodeURIComponent(repo)}/compare/${base}...${head}`,
    {
      headers: headers(token),
      timeout: 20000,
      params: { per_page: 100 },
    },
  );

  return response.data.commits || [];
}


export async function repositoryEvents(guildId, owner, repo) {
  const token = await guildToken(guildId);

  const response = await axios.get(
    `${API}/repos/${encodeURIComponent(owner)}/${encodeURIComponent(repo)}/events`,
    {
      headers: headers(token),
      timeout: 20000,
      params: { per_page: 100 },
    },
  );

  return response.data || [];
}
